In [1]:
import numpy
import scipy.special #sigmoid function expit()
import matplotlib.pyplot #plotting arrays
%matplotlib inline

In [2]:
#definição da classe Rede Neural
class NeuralNet:
    
    #inicializar a rede
    def __init__(self, inputN, hiddenN, outputN, learningR):
        
        #Configurar o numero de nós em cada camada
        self.inodes = inputN
        self.hnodes = hiddenN
        self.onodes = outputN
        
        #taxa de aprendizado
        self.lr = learningR
        
        #WEIGHTS
        #Random weights from a normal probability distribution centred on zero
        #standard deviation = 1/srqt(incoming links)
        
        #numpy.random.normal(mean, spread, shape)
        self.wih = numpy.random.normal(0.0, pow(self.inodes, -0.5), (self.hnodes, self.inodes))
        self.who = numpy.random.normal(0.0, pow(self.hnodes, -0.5), (self.onodes, self.hnodes))
        
        #ACTIVATION FUNCTION
        self.activation_function = lambda x : scipy.special.expit(x)
        
        pass
    
    #treinar a rede
    def train(self, inputs_list, targets_list):
        #convertere para array2d numpy
        inputs = numpy.array(inputs_list, ndmin=2).T
        targets = numpy.array(targets_list, ndmin=2).T
        
        #PROPAGATION
        #signals hiddenl
        hi = numpy.dot(self.wih, inputs) #hidden inputs
        ho = self.activation_function(hi) #hidden outputs
        
        #signals outputL
        oi = numpy.dot(self.who, ho) #final inputs
        oo = self.activation_function(oi) #final outputs
        
        #BACKpropagation
        #ERRORS
        output_errors = targets - oo #error = (target - actual)
        hidden_errors = numpy.dot(self.who.T, output_errors) #ErrorsH = who.T dot errorsOut
        
        #UPDATING WEIGHTS
        self.who += self.lr * numpy.dot((output_errors * oo * (1.0 - oo)), numpy.transpose(ho))
        self.wih += self.lr * numpy.dot((hidden_errors * ho * (1.0 - ho)), numpy.transpose(inputs))
        pass
    
    #resposta depois de treinada
    def query(self, inputs_list):
        
        #Convertendo inputs para Array2d
        inputs = numpy.array(inputs_list, ndmin=2).T
        
        #INPUT --> HIDDEN
        #Hidden input calculation
        hidden_inputs = numpy.dot(self.wih, inputs)
        
        #Hidden outputs (activation function)
        hidden_outputs = self.activation_function(hidden_inputs)
        
        #HIDDEN --> OUTPUT
        #signals into the final layer (output layer)
        output_inputs = numpy.dot(self.who, hidden_outputs)
        #signals out the final layer (output layer)
        output_outputs = self.activation_function(output_inputs)
        
        return output_outputs

In [3]:
#MNIST NUMBERS Creating instance of the NN
#INPUTS (Every pixel 28x28), hidden, output (10 possible numbers)
inputl = 784
hiddenl = 300
outl = 10

learningrate = 0.1

nn = NeuralNet(inputl, hiddenl, outl, learningrate)

In [4]:
#LOADING THE DATA
training_data_arq = open("mnist/mnist_train.csv", "r")
training_data_lista = training_data_arq.readlines()
training_data_arq.close()

In [5]:
#PREPARING THE DATA and TRAINING THE NN
# epochs is the number of times the training dataset is used for training
epochs = 5
for i in range(epochs):
    for record in training_data_lista:
        all_values = record.split(',') #separando a string atual pelas virgulas
        inputs = (numpy.asfarray(all_values[1:]) / 255.0 * 0.99) + 0.01 #scaling the inputs
        target = numpy.zeros(outl) + 0.01 #criando o target 0.01-0.99
        target[int(all_values[0])] = 0.99 #target label

        nn.train(inputs, target)
        pass
    pass

In [6]:
# --TESTING THE NETWORK--
#loading the mnist testing data
test_df = open("mnist/mnist_test.csv", "r")
test_l = test_df.readlines()
test_df.close()

In [7]:
scorecard = []
for record in test_l:
    all_values = record.split(',')
    inputs = (numpy.asfarray(all_values[1:]) / 255.00 * 0.99) + 0.01
    
    predicted = numpy.argmax(nn.query(inputs)) #predicted value
    
    #scoreboard
    if (predicted == int(all_values[0])):
        scorecard.append(1)
    else:
        scorecard.append(0)
    pass

In [8]:
#calculating performance
scorecard_array = numpy.asarray(scorecard)
performance = scorecard_array.sum() / scorecard_array.size
print ("Performance =",  performance)
print ("Error =",  1.0 - performance)

Performance = 0.9744
Error = 0.025599999999999956


**100 hidden layers**

Performance = 0.9453 0.3
Performance = 0.9465 0.3

Performance = 0.95 <- look at this. 0.950000000000 0.2
Error = 0.050000000000000044

epochs
Performance = 0.9566 lr 0.2 e 2
Error = 0.043399999999999994

Performance = 0.9588 lr 0.2 e 2
Error = 0.041200000000000014

Performance = 0.9636 lr 0.2 e 5
Error = 0.03639999999999999

Performance = 0.9686 lr 0.1 e 5
Error = 0.031399999999999983

**Increase hidden nodes to 300**

Performance = 0.9744 lr 0.1 e 5
Error = 0.025599999999999956